pkg/runner/test_runner_route_test.go

```diff
@@ -0,0 +1,403 @@
+		require.NoError(suite.T(), err)
+	}
+	ts1, err := pluginRegistry.NewTestStep("Example")
+	require.NoError(suite.T(), err)
+
+	params := make(test.TestStepParameters)
+	bundle := test.TestStepBundle{TestStep: ts1, TestStepLabel: "FirstStage", Parameters: params}
+
+	routeInCh := make(chan *target.Target)
+	routeOutCh := make(chan *target.Target)
+	stepInCh := make(chan *target.Target)
+	stepOutCh := make(chan *target.Target)
+	stepErrCh := make(chan cerrors.TargetError)
+	targetErrCh := make(chan cerrors.TargetError)
+
+	suite.routeInCh = routeInCh
+	suite.routeOutCh = routeOutCh
+	suite.stepOutCh = stepOutCh
+	suite.stepInCh = stepInCh
+	suite.stepErrCh = stepErrCh
+	suite.targetErrCh = targetErrCh
+
+	suite.routingChannels = routingCh{
+		routeIn:   routeInCh,
+		routeOut:  routeOutCh,
+		stepIn:    stepInCh,
+		stepErr:   stepErrCh,
+		stepOut:   stepOutCh,
+		targetErr: targetErrCh,
+	}
+
+	s, err := memory.New()
+	require.NoError(suite.T(), err)
+	storage.SetStorage(s)
+
+	header := testevent.Header{
+		JobID:         1,
+		RunID:         1,
+		TestName:      "TestRunnerSuite",
+		TestStepLabel: "TestRunnerSuite",
+	}
+	ev := storage.NewTestEventEmitterFetcher(header)
+
+	timeouts := TestRunnerTimeouts{
+		StepInjectTimeout:   30 * time.Second,
+		MessageTimeout:      5 * time.Second,
+		ShutdownTimeout:     1 * time.Second,
+		StepShutdownTimeout: 1 * time.Second,
+	}
+
+	suite.router = newRouter(log, bundle, suite.routingChannels, ev, timeouts)
+}
+
+func (suite *TestRunnerSuite) TestRouteInRoutesAllTargets() {
+
+	// test that all targets are routed to step input channel without delay, in order
+	targets := []*target.Target{
+		&target.Target{Name: "host001", ID: "001", FQDN: "host001.facebook.com"},
+		&target.Target{Name: "host002", ID: "002", FQDN: "host002.facebook.com"},
+		&target.Target{Name: "host003", ID: "003", FQDN: "host003.facebook.com"},
+	}
+
+	routeInResult := make(chan error)
+	stepInResult := make(chan error)
+	terminate := make(chan struct{})
+
+	go func() {
+		// start routing
+		_, _ = suite.router.routeIn(terminate)
+	}()
+
+	// inject targets
+	go func() {
+		defer close(suite.routeInCh)
+		for _, target := range targets {
+			select {
+			case <-time.After(2 * time.Second):
+				routeInResult <- fmt.Errorf("target should be accepted by routeIn block within timeout")
+				return
+			case suite.routeInCh <- target:
+			}
+		}
+		routeInResult <- nil
+	}()
+
+	// mimic the test step consuming targets in input
+	go func() {
+		numTargets := 0
+		for {
+			select {
+			case t, ok := <-suite.stepInCh:
+				if !ok {
+					if numTargets != len(targets) {
+						stepInResult <- fmt.Errorf("not all targets received by teste step")
+					} else {
+						stepInResult <- nil
+					}
+					return
+				}
+				if numTargets+1 > len(targets) {
+					stepInResult <- fmt.Errorf("more targets returned than injected")
+					return
+				}
+				if t.Name != targets[numTargets].Name || t.ID != targets[numTargets].ID {
```

Comment:
Should it be the same pointer? Why we compare fields (which could change in future) if we can compare pointers?

Reply:
Ok, so this is a long standing problem. What makes two targets equal? So that at any point in time in a test there will not be two targets that are considered equal? There should be only one place where this is defined, e.g. `t1.Equal(t2)`, and in my opinion pointer comparison is not the right approach. However, pointer comparison is used basically all over the place in the runner, so I'll move to pointer comparison for consistency, but I'd love if we moved soon to a more robust mechanism.

Reply:
Yeah, OK. This is a subject for discussion what internal guarantees we provide. And before that your approach is safer, so let's leave it as is, agreed.

Reply:
Just for the record, I reverted to pointer comparison for consistency. I am in favor of establishing the single mechanism mentioned above, which we agree provides better guarantees, but I'd rather be consistent here with what the rest of the code (and actually the pipeline which is being unit test here, does).